File: gc-app/src/main/java/rocks/frick/gc/application/UploadActivityService.java
package rocks.frick.gc.application;

import java.io.InputStream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import rocks.frick.gc.Ride;
import rocks.frick.gc.WorkoutAnalyzer;
import rocks.frick.gc.WorkoutSummary;
import rocks.frick.gc.application.spi.ActivityRepository;

public class UploadActivityService {

  private static final Logger log = LoggerFactory.getLogger(UploadActivityService.class);

  private final ActivityRepository activityRepository;
  private final WorkoutAnalyzer analyzer;

  public UploadActivityService(ActivityRepository activityRepository, WorkoutAnalyzer analyzer) {
    this.activityRepository = activityRepository;
    this.analyzer = analyzer;
  }

  public String upload(InputStream data, String name) {
    WorkoutSummary summary = analyzer.analyze(data);

    log.debug("{}", summary);
    var ride = new Ride(
      summary.ride().id(),
      summary.ride().samples(),
      name,
      summary.ride().source(),
      summary.ride().file_id(),
      summary.ride().user_profile()
    );

    return activityRepository.save(ride);

  }
}

File: gc-core/src/main/scala/gc/application/UploadActivityService.scala
package gc.application
import scala.jdk.CollectionConverters._

import gc.WorkoutAnalyzer
import gc.spi.SaveActivity
import com.typesafe.scalalogging.Logger
import java.io.InputStream
import gc.model.Activity

class UploadActivityService(repo: SaveActivity, analyzer: WorkoutAnalyzer) {
  private val log = Logger("UploadActivityService")

  def upload(data: InputStream, name: String): Activity = {
    val summary = analyzer.analyze(data)

      // TODO: kill the Java types
    val ride = Activity(
      id           =  summary.activity.id,
      name         =  name,
      samples      =  summary.activity.samples,
      source       =  summary.source,
      fileID       =  summary.activity.fileID,
      userProfile  =  summary.activity.userProfile
    )
    repo.save(ride)
  }

}
